File: 01Algorithm_Lecture/BaaaaaaakingDog_Algoritm/cha0x09_BOJ_2667.py
def bfs(n,sx,sy,myMap,visit,danji):

    queue = [[sx,sy]]
    cnt = 1
    dx = [0,0,-1,1]
    dy = [1,-1,0,0]
    while(queue):
        x,y = queue.pop(0)
        visit[x][y] = danji
        for i in range(4):
            cx = x + dx[i]
            cy = y + dy[i]
            if(cx<0 or cy<0 or cx >=n or cy >=n):
                continue
            else:
                if(myMap[cx][cy] == 1 and visit[cx][cy] == 0):
                    visit[cx][cy] = danji
                    queue.append([cx,cy])
                    cnt +=1

    return visit,cnt

n = int(input())

myMap = []

for i in range(n):
    myMap.append(list(map(int,input())))

visit=[[0 for i in range(n)]for j in range(n)]

danji = 0
list_danji = []
for i in range(n):
    for j in range(n):
        if(myMap[i][j] == 1 and visit[i][j] == 0):
            danji +=1
            visit,cnt = bfs(n,i,j,myMap,visit,danji)
            list_danji.append(cnt)

list_danji.sort()
print(len(list_danji))
for i in range(len(list_danji)):
    print(list_danji[i])

File: 01Algorithm_Lecture/BaaaaaaakingDog_Algoritm/cha0x09_BOJ_2667.cpp
#include <iostream>
#include <queue>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;
int myMap[26][26];
int visit[26][26];
int dx[4] = {0,0,-1,1};
int dy[4] = {1,-1,0,0};
vector<int> cntDanji={0};
int n;

void bfs(int sx, int sy, int danji){

    queue <int> q;
    q.push(sx);
    q.push(sy);
    visit[sx][sy] = danji;
    int cnt = 1;
    while(!q.empty()){
        int x = q.front();
        q.pop();
        int y = q.front();
        q.pop();
        for(int dir = 0; dir< 4; dir++){
            int cx = x + dx[dir];
            int cy = y + dy[dir];
            if(cx<0||cy<0||cx>=n||cy>=n){
                continue;
            }
            else{
                if(myMap[cx][cy] == 1 && visit[cx][cy] == 0){
                    q.push(cx);
                    q.push(cy);
                    visit[cx][cy] = danji;
                    cnt ++;
                }
            }
        }

    }
    cntDanji.push_back(cnt);
}


int main(){

    ios::sync_with_stdio(0);
    cin.tie(0);
    string tmp;
    cin >> n;
    for(int i = 0; i < n; i++){
        cin >> tmp;
        for(int j = 0; j < n ; j++){
            myMap[i][j] = tmp[j] - '0';
        }
    }
    int danji = 0;
    for(int i = 0; i < n ; i ++){
        for(int j = 0; j < n; j++){
            if(myMap[i][j] == 1 && visit[i][j] == 0){
                danji ++;
                
                bfs(i, j, danji);
            }
        }
    }

    sort(cntDanji.begin(),cntDanji.end());
    cout<<danji << '\n';
    for(int i = 1 ; i < cntDanji.size() ;i++){
        cout << cntDanji[i]<<'\n';
    }

}